cxx/util_io.cc

```diff
@@ -175,26 +177,41 @@ T_observations load_observations(const std::string& path, T_schema& schema) {
     std::string relname;
     std::vector<std::string> items;
 
-    stream >> value_str;
-    stream >> relname;
-    ObservationVariant value = observation_string_to_value(
-        value_str,
-        std::visit(
-            [](const auto& trel) {
-              using T = std::decay_t<decltype(trel)>;
-              if constexpr (std::is_same_v<T, T_clean_relation>) {
-                return trel.distribution_spec.observation_type;
-              } else if constexpr (std::is_same_v<T, T_noisy_relation>) {
-                return trel.emission_spec.observation_type;
-              } else {
-                assert(false && "Unrecognized relation type.");
-              }
-            },
-            schema.at(relname)));
-    for (std::string w; stream >> w;) {
-      items.push_back(w);
+    if (!getline(stream, value_str, ',')) {
+      assert(false && "Error parsing schema. No comma separating value.");
     }
-    assert(items.size() > 0);
+
+    if (!getline(stream, relname, ',')) {
+      assert(false && "Error parsing schema. No comma separating relation.");
+    }
+
+    ObservationVariant value;
+
+    try {
+      value = observation_string_to_value(
+          value_str,
+          std::visit(
+              [](const auto& trel) {
+                using T = std::decay_t<decltype(trel)>;
+                if constexpr (std::is_same_v<T, T_clean_relation>) {
+                  return trel.distribution_spec.observation_type;
+                } else if constexpr (std::is_same_v<T, T_noisy_relation>) {
+                  return trel.emission_spec.observation_type;
+                } else {
+                  assert(false && "Unrecognized relation type.");
+                }
+              },
+              schema.at(relname)));
+    } catch(const std::out_of_range& e) {
```

Comment:
Updated. Sticking with assert seems reasonable, since that is Google compatible. Do you mean using abseil Status and passing that around (to be Google compatible)?

Reply:
I was thinking abseil Status, but I haven't done much thinking/investigating other options (there's `std::expected`, but that's not until C++23)